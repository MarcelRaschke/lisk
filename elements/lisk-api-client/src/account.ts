/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { codec } from '@liskhq/lisk-codec';
import { Channel, RegisteredSchemas } from './types';

export class Account {
	private readonly _channel: Channel;
	private readonly _schemas: RegisteredSchemas;

	public constructor(channel: Channel, schemas: RegisteredSchemas) {
		this._channel = channel;
		this._schemas = schemas;
	}

	public async get(address: Buffer | string): Promise<Record<string, unknown>> {
		const addressString: string = Buffer.isBuffer(address) ? address.toString('hex') : address;
		const accountHex: string = await this._channel.invoke('app:getAccount', {
			address: addressString,
		});

		return this.decode(Buffer.from(accountHex, 'hex'));
	}

	public encode(input: Record<string, unknown>): Buffer {
		return codec.encode(this._schemas.account, input);
	}

	public decode(input: Buffer | string): Record<string, unknown> {
		const inputBuffer: Buffer = Buffer.isBuffer(input) ? input : Buffer.from(input, 'hex');
		return codec.decode(this._schemas.account, inputBuffer);
	}

	public toJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.toJSON(this._schemas.account, account);
	}

	public fromJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.fromJSON(this._schemas.account, account);
	}
}
