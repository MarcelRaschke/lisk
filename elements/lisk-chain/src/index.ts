/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK, EVENT_VALIDATORS_CHANGED } from './constants';

const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK, EVENT_VALIDATORS_CHANGED };

export { events };
export { Chain } from './chain';
export { Transaction, TransactionInput, transactionSchema, calculateMinFee } from './transaction';
export {
	blockHeaderAssetSchema,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	validatorsSchema,
	getGenesisBlockHeaderAssetSchema,
	stateDiffSchema,
	getRegisteredBlockAssetSchema,
} from './schema';
export {
	CONSENSUS_STATE_VALIDATORS_KEY,
	CONSENSUS_STATE_FINALIZED_HEIGHT_KEY,
	TAG_BLOCK_HEADER,
	TAG_TRANSACTION,
} from './constants';
export type {
	Account,
	AccountDefaultProps,
	RawBlock,
	RawBlockHeader,
	GenesisBlock,
	GenesisBlockHeader,
	Block,
	BlockHeader,
	BlockHeaderAsset,
	Validator,
	AccountSchema,
} from './types';
export { Slots } from './slots';
export { readGenesisBlockJSON, getValidators, getAccountSchemaWithDefault } from './utils';
export * as testing from './testing';

export type { StateStore } from './state_store';
export type { DataAccess } from './data_access';
