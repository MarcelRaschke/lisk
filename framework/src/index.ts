/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export {
	Transaction,
	transactionSchema,
	readGenesisBlockJSON,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	getAccountSchemaWithDefault,
	getGenesisBlockHeaderAssetSchema,
	blockHeaderAssetSchema,
} from '@liskhq/lisk-chain';
export {
	BaseModule,
	BaseAsset,
	TokenModule,
	TokenTransferAsset,
	SequenceModule,
	KeysModule,
	KeysRegisterAsset,
	DPoSModule,
	DPoSRegisterAsset,
	DPoSVoteAsset,
	DPoSUnlockAsset,
	DPoSPoMAsset,
} from './modules';
export { Application } from './application';
export { systemDirs } from './system_dirs';
export { BasePlugin, BlockHeaderJSON, PluginInfo, PluginCodec } from './plugins/base_plugin';
export { IPCChannel } from './controller/channels';
export type { BaseChannel } from './controller/channels';
export type { EventsDefinition, EventCallback } from './controller/event';
export type { ActionsDefinition, ActionHandler } from './controller/action';
export * as testing from './testing';
export * from './types';
