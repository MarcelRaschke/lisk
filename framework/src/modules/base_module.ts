/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable class-methods-use-this */

import {
	GenesisConfig,
	AccountSchema,
	TransactionApplyContext,
	AfterBlockApplyContext,
	BeforeBlockApplyContext,
	AfterGenesisBlockApplyContext,
	Reducers,
	Actions,
	BaseModuleDataAccess,
} from '../types';
import { BaseAsset } from './base_asset';
import { Logger } from '../logger/logger';

export interface BaseModuleChannel {
	publish(name: string, data?: Record<string, unknown>): void;
}

export abstract class BaseModule {
	public readonly config: GenesisConfig;
	public transactionAssets: BaseAsset[] = [];
	public reducers: Reducers = {};
	public actions: Actions = {};
	public events: string[] = [];
	public accountSchema?: AccountSchema;
	protected _logger!: Logger;
	protected _channel!: BaseModuleChannel;
	protected _dataAccess!: BaseModuleDataAccess;
	public abstract name: string;
	public abstract id: number;

	public constructor(genesisConfig: GenesisConfig) {
		this.config = genesisConfig;
	}

	public init(input: {
		channel: BaseModuleChannel;
		dataAccess: BaseModuleDataAccess;
		logger: Logger;
	}): void {
		this._channel = input.channel;
		this._dataAccess = input.dataAccess;
		this._logger = input.logger;
	}

	public async beforeTransactionApply?(context: TransactionApplyContext): Promise<void>;
	public async afterTransactionApply?(context: TransactionApplyContext): Promise<void>;
	public async afterGenesisBlockApply?(context: AfterGenesisBlockApplyContext): Promise<void>;
	public async beforeBlockApply?(context: BeforeBlockApplyContext): Promise<void>;
	public async afterBlockApply?(context: AfterBlockApplyContext): Promise<void>;
}
