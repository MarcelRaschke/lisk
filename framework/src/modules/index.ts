/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export * from './base_module';
export * from './base_asset';
export { KeysModule, RegisterAsset as KeysRegisterAsset } from './keys';
export { TokenModule, TransferAsset as TokenTransferAsset } from './token';
export { SequenceModule } from './sequence';
export {
	DPoSModule,
	RegisterTransactionAsset as DPoSRegisterAsset,
	VoteTransactionAsset as DPoSVoteAsset,
	UnlockTransactionAsset as DPoSUnlockAsset,
	PomTransactionAsset as DPoSPoMAsset,
} from './dpos';
