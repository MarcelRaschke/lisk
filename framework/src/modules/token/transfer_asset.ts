/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable class-methods-use-this */
import { MAX_TRANSACTION_AMOUNT } from './constants';
import { TokenAccount, Asset } from './types';
import { BaseAsset } from '../base_asset';
import { ApplyAssetContext } from '../../types';

export class TransferAsset extends BaseAsset {
	public name = 'transfer';
	public id = 0;
	public schema = {
		$id: 'lisk/transfer-asset',
		title: 'Transfer transaction asset',
		type: 'object',
		required: ['amount', 'recipientAddress', 'data'],
		properties: {
			amount: {
				dataType: 'uint64',
				fieldNumber: 1,
			},
			recipientAddress: {
				dataType: 'bytes',
				fieldNumber: 2,
				minLength: 20,
				maxLength: 20,
			},
			data: {
				dataType: 'string',
				fieldNumber: 3,
				minLength: 0,
				maxLength: 64,
			},
		},
	};
	private readonly _minRemainingBalance: bigint;

	public constructor(minRemainingBalance: bigint) {
		super();
		this._minRemainingBalance = minRemainingBalance;
	}

	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<Asset>): Promise<void> {
		const sender = await stateStore.account.get<TokenAccount>(transaction.senderAddress);
		if (!sender) {
			throw new Error(
				`Account does not exist for senderAddress: ${transaction.senderAddress.toString('hex')}`,
			);
		}
		sender.token.balance -= asset.amount;
		await stateStore.account.set(sender.address, sender);
		const recipient = await stateStore.account.getOrDefault<TokenAccount>(asset.recipientAddress);
		recipient.token.balance += asset.amount;
		if (recipient.token.balance > BigInt(MAX_TRANSACTION_AMOUNT)) {
			throw new Error(
				`Invalid transfer amount: ${asset.amount.toString()}. Maximum allowed balance for recipient is: ${MAX_TRANSACTION_AMOUNT}`,
			);
		}

		if (recipient.token.balance < this._minRemainingBalance) {
			throw new Error(
				`Recipient account ${recipient.address.toString(
					'hex',
				)} does not meet the minimum remaining balance requirement: ${this._minRemainingBalance}.`,
			);
		}

		await stateStore.account.set(recipient.address, recipient);
	}
}
