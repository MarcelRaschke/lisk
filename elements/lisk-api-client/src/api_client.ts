/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { EventCallback, Channel, RegisteredSchemas, NodeInfo } from './types';
import { Node } from './node';
import { Account } from './account';
import { Block } from './block';
import { Transaction } from './transaction';

export class APIClient {
	private readonly _channel: Channel;
	private _schemas!: RegisteredSchemas;
	private _nodeInfo!: NodeInfo;
	private _node!: Node;
	private _account!: Account;
	private _block!: Block;
	private _transaction!: Transaction;

	public constructor(channel: Channel) {
		this._channel = channel;
	}

	public async init(): Promise<void> {
		this._schemas = await this._channel.invoke<RegisteredSchemas>('app:getSchema');
		this._node = new Node(this._channel);
		this._account = new Account(this._channel, this._schemas);
		this._block = new Block(this._channel, this._schemas);
		this._nodeInfo = await this._node.getNodeInfo();
		this._transaction = new Transaction(this._channel, this._schemas, this._nodeInfo);
	}

	public async disconnect(): Promise<void> {
		return this._channel.disconnect();
	}

	public async invoke<T = Record<string, unknown>>(
		actionName: string,
		params?: Record<string, unknown>,
	): Promise<T> {
		return this._channel.invoke(actionName, params);
	}

	public subscribe(eventName: string, cb: EventCallback): void {
		this._channel.subscribe(eventName, cb);
	}

	public get schemas(): RegisteredSchemas {
		return this._schemas;
	}

	public get node(): Node {
		return this._node;
	}

	public get account(): Account {
		return this._account;
	}

	public get block(): Block {
		return this._block;
	}

	public get transaction(): Transaction {
		return this._transaction;
	}
}
