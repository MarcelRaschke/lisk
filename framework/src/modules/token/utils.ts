/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Block } from '@liskhq/lisk-chain';

export const getTotalFees = (
	block: Block,
	minFeePerByte: bigint,
	baseFees: ReadonlyArray<{ assetID: number; baseFee: string; moduleID: number }>,
): { readonly totalFee: bigint; readonly totalMinFee: bigint } =>
	block.payload.reduce(
		(prev, current) => {
			const baseFee =
				baseFees.find(
					(fee: { moduleID: number; assetID: number }) =>
						fee.moduleID === current.moduleID && fee.assetID === current.assetID,
				)?.baseFee ?? BigInt(0);
			const minFee = minFeePerByte * BigInt(current.getBytes().length) + BigInt(baseFee);

			return {
				totalFee: prev.totalFee + current.fee,
				totalMinFee: prev.totalMinFee + minFee,
			};
		},
		{ totalFee: BigInt(0), totalMinFee: BigInt(0) },
	);
