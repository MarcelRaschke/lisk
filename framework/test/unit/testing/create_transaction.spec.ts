/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { TransferAsset } from '../../../src/modules/token';
import { createTransaction } from '../../../src/testing';

describe('Create Transaction', () => {
	const asset = {
		amount: BigInt('1000000000000'),
		recipientAddress: Buffer.from('ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815', 'hex'),
		data: 'moon',
	};

	it('should return a valid transaction', () => {
		expect(createTransaction({ moduleID: 2, assetClass: TransferAsset, asset })).toMatchSnapshot();
	});

	it('should return valid signed transaction with passphrase', () => {
		const transaction = createTransaction({
			moduleID: 2,
			assetClass: TransferAsset,
			asset,
			passphrase: 'pass',
			networkIdentifier: Buffer.alloc(1),
		});

		expect(transaction.signatures).toHaveLength(1);
	});
});
