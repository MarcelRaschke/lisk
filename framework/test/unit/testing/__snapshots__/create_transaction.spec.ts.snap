// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Transaction should return a valid transaction 1`] = `
Transaction {
  "asset": Object {
    "data": Array [
      8,
      128,
      160,
      148,
      165,
      141,
      29,
      18,
      20,
      171,
      0,
      65,
      167,
      211,
      247,
      178,
      194,
      144,
      181,
      184,
      52,
      212,
      107,
      220,
      123,
      126,
      184,
      88,
      21,
      26,
      4,
      109,
      111,
      111,
      110,
    ],
    "type": "Buffer",
  },
  "assetID": 0,
  "fee": 0n,
  "moduleID": 2,
  "nonce": 0n,
  "senderPublicKey": Object {
    "data": Array [
      190,
      144,
      123,
      75,
      172,
      132,
      254,
      229,
      206,
      136,
      17,
      219,
      45,
      239,
      201,
      191,
      11,
      42,
      42,
      43,
      188,
      61,
      84,
      216,
      162,
      37,
      126,
      205,
      112,
      68,
      25,
      98,
    ],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;
